File: lib/features/assessment/domain/frequency_behavior_v2/frequency_behavior_v2_selector.dart
import 'frequency_behavior_v2_contract.dart';
import 'frequency_behavior_v2_models.dart';
import 'frequency_behavior_v2_rng.dart';

/// Dormant 50-of-405 Frequency V2 session composer.
///
/// Not wired to live Frequency routing. Does not read age, profession,
/// location, gender, previous answers, personality estimates, behavioral
/// vectors, or response speed. Evidence scores and weight magnitudes are
/// not pick keys. Semantic-cluster mix is a bounded lookahead preference,
/// not a hard quota.
class FrequencyBehaviorV2SessionComposer {
  const FrequencyBehaviorV2SessionComposer();

  static const int perDimensionQuota =
      FrequencyBehaviorV2Contract.sessionBasePerDimension;
  static const int flexSlots = FrequencyBehaviorV2Contract.sessionFlexSlots;

  List<FrequencyBehaviorV2SessionItemPlan> compose({
    required FrequencyBehaviorV2PoolDocument pool,
    required String sessionSeed,
    Map<String, Map<String, dynamic>> reviewByItemId = const {},
    List<List<String>> nearDuplicateClusters = const [],
    bool excludeUnresolvedReview = true,
    String? sessionId,
    String? createdAt,
  }) {
    return composeManifest(
      pool: pool,
      sessionSeed: sessionSeed,
      reviewByItemId: reviewByItemId,
      nearDuplicateClusters: nearDuplicateClusters,
      excludeUnresolvedReview: excludeUnresolvedReview,
      sessionId: sessionId,
      createdAt: createdAt,
    ).itemPlans;
  }

  FrequencyBehaviorV2SessionManifest composeManifest({
    required FrequencyBehaviorV2PoolDocument pool,
    required String sessionSeed,
    Map<String, Map<String, dynamic>> reviewByItemId = const {},
    List<List<String>> nearDuplicateClusters = const [],
    bool excludeUnresolvedReview = true,
    String? sessionId,
    String? createdAt,
  }) {
    final selectorVersion = FrequencyBehaviorV2Contract.selectorVersion;
    final bankVersion = pool.poolVersion;
    final eligible = <FrequencyBehaviorV2Item>[
      for (final item in pool.items)
        if (!excludeUnresolvedReview || _isSelectable(item, reviewByItemId))
          item,
    ];
    final byId = {for (final i in eligible) i.itemId: i};
    final nearDupIndex =
        _nearDupIndex(nearDuplicateClusters, byId.keys.toSet());

    final extraRng = FrequencyBehaviorV2Rng.forStream(
      selectorVersion: selectorVersion,
      bankVersion: bankVersion,
      sessionSeed: sessionSeed,
      stream: 'extra_slots',
    );
    final extraOrder = extraRng.shuffledCopy(
      FrequencyBehaviorV2Contract.canonicalDimensions,
    );
    final extraDims = <String>{extraOrder[0], extraOrder[1]};

    final pickedByDim = <String, List<FrequencyBehaviorV2Item>>{};
    final usedNearDup = <int>{};

    for (final dim in FrequencyBehaviorV2Contract.canonicalDimensions) {
      final quota =
          extraDims.contains(dim) ? perDimensionQuota + 1 : perDimensionQuota;
      final ranked = [
        for (final item in eligible)
          if (item.primaryDimensions.single == dim) item,
      ];
      ranked.sort((a, b) {
        final ra = _candidateRank(
          selectorVersion: selectorVersion,
          bankVersion: bankVersion,
          sessionSeed: sessionSeed,
          dimension: dim,
          questionId: a.itemId,
        );
        final rb = _candidateRank(
          selectorVersion: selectorVersion,
          bankVersion: bankVersion,
          sessionSeed: sessionSeed,
          dimension: dim,
          questionId: b.itemId,
        );
        if (ra != rb) return ra.compareTo(rb);
        return a.itemId.compareTo(b.itemId);
      });
      if (ranked.length < quota) {
        throw StateError(
          'insufficient_candidates:$dim have=${ranked.length} need=$quota',
        );
      }
      final lookahead = FrequencyBehaviorV2Contract.softClusterLookahead;
      final picked = <FrequencyBehaviorV2Item>[];
      final clusterCounts = <String, int>{};
      final pickedSet = <String>{};

      bool nearDupBlocked(FrequencyBehaviorV2Item c) {
        final nd = nearDupIndex[c.itemId];
        return nd != null && usedNearDup.contains(nd);
      }

      void absorb(FrequencyBehaviorV2Item c) {
        pickedSet.add(c.itemId);
        picked.add(c);
        clusterCounts[c.semanticCluster] =
            (clusterCounts[c.semanticCluster] ?? 0) + 1;
        final nd = nearDupIndex[c.itemId];
        if (nd != null) usedNearDup.add(nd);
      }

      // Seed rank is primary. Repeating a cluster only prefers another
      // unused cluster inside a short lookahead; it never scans the bank
      // to satisfy a diversity quota.
      while (picked.length < quota) {
        final remaining = [
          for (final c in ranked)
            if (!pickedSet.contains(c.itemId) && !nearDupBlocked(c)) c,
        ];
        if (remaining.isEmpty) {
          throw StateError(
            'underfilled:$dim have=${picked.length} need=$quota',
          );
        }
        final head = remaining.first;
        var chosen = head;
        if ((clusterCounts[head.semanticCluster] ?? 0) > 0) {
          final altLimit = lookahead < remaining.length - 1
              ? lookahead
              : remaining.length - 1;
          for (var i = 1; i <= altLimit; i++) {
            final alt = remaining[i];
            if ((clusterCounts[alt.semanticCluster] ?? 0) == 0) {
              chosen = alt;
              break;
            }
          }
        }
        absorb(chosen);
      }
      pickedByDim[dim] = picked;
    }

    final queues = <String, List<FrequencyBehaviorV2Item>>{};
    for (final dim in FrequencyBehaviorV2Contract.canonicalDimensions) {
      final qRng = FrequencyBehaviorV2Rng.forStream(
        selectorVersion: selectorVersion,
        bankVersion: bankVersion,
        sessionSeed: sessionSeed,
        stream: 'queue|$dim',
      );
      queues[dim] = qRng.shuffledCopy(pickedByDim[dim]!);
    }

    final interleaveRng = FrequencyBehaviorV2Rng.forStream(
      selectorVersion: selectorVersion,
      bankVersion: bankVersion,
      sessionSeed: sessionSeed,
      stream: 'interleave_order',
    );
    final dimOrder = interleaveRng.shuffledCopy(
      FrequencyBehaviorV2Contract.canonicalDimensions,
    );

    final sequence = <FrequencyBehaviorV2Item>[];
    for (var round = 0; round < perDimensionQuota; round++) {
      for (final dim in dimOrder) {
        sequence.add(queues[dim]!.removeAt(0));
      }
    }
    final extras = <FrequencyBehaviorV2Item>[];
    for (final dim in dimOrder) {
      extras.addAll(queues[dim]!);
    }
    for (final extra in extras) {
      _insertExtra(sequence, extra);
    }
    _softenAdjacentFamilies(sequence);

    if (sequence.length != FrequencyBehaviorV2Contract.sessionItemCount) {
      throw StateError('session_length_${sequence.length}');
    }
    if (!_consecutivePrimaryOk(sequence)) {
      throw StateError('consecutive_primary_violation');
    }

    final derivedId = sessionId ??
        'frequency_v2_${FrequencyBehaviorV2Rng.fnv1a32('$selectorVersion|$bankVersion|$sessionSeed').toRadixString(16).padLeft(8, '0')}';

    final questions = <FrequencyBehaviorV2SessionQuestion>[];
    for (var i = 0; i < sequence.length; i++) {
      final item = sequence[i];
      final optRng = FrequencyBehaviorV2Rng.forStream(
        selectorVersion: selectorVersion,
        bankVersion: bankVersion,
        sessionSeed: sessionSeed,
        stream: 'options|${item.itemId}',
      );
      final authored = [for (final o in item.options) o.optionId];
      final presented = optRng.shuffledCopy(authored);
      questions.add(
        FrequencyBehaviorV2SessionQuestion(
          questionId: item.itemId,
          primaryDimension: item.primaryDimensions.single,
          presentationIndex: i,
          presentedOptionOrder: presented,
        ),
      );
    }

    return FrequencyBehaviorV2SessionManifest(
      schemaVersion: FrequencyBehaviorV2Contract.sessionManifestSchemaVersion,
      selectorVersion: selectorVersion,
      bankVersion: bankVersion,
      sessionId: derivedId,
      sessionSeed: sessionSeed,
      locale: pool.locale,
      createdAt: createdAt,
      questionIds: [for (final q in questions) q.questionId],
      questions: questions,
    );
  }

  static int _candidateRank({
    required String selectorVersion,
    required String bankVersion,
    required String sessionSeed,
    required String dimension,
    required String questionId,
  }) {
    return FrequencyBehaviorV2Rng.fromParts([
      selectorVersion,
      bankVersion,
      sessionSeed,
      dimension,
      questionId,
    ]).nextUint32();
  }

  static Map<String, int> _nearDupIndex(
    List<List<String>> clusters,
    Set<String> eligibleIds,
  ) {
    final out = <String, int>{};
    for (var i = 0; i < clusters.length; i++) {
      final members = [
        for (final id in clusters[i])
          if (eligibleIds.contains(id)) id,
      ];
      if (members.length < 2) continue;
      for (final id in members) {
        out.putIfAbsent(id, () => i);
      }
    }
    return out;
  }

  static bool _keyGreater(List<int> a, List<int> b) {
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return a[i] > b[i];
    }
    return false;
  }

  static String _contextFamily(FrequencyBehaviorV2Item item) {
    if (item.context.isEmpty) return 'unclassified';
    return item.context.first;
  }

  static String? _thematicTag(FrequencyBehaviorV2Item item) {
    final text = '${item.prompt} ${item.semanticCluster}'.toLowerCase();
    if (RegExp(r'aile|anne|baba|kayın|çocuk|cocuk').hasMatch(text)) {
      return 'family';
    }
    if (RegExp(r'para|fatura|hesap|ödeme|odeme|borç|borc').hasMatch(text)) {
      return 'payment';
    }
    if (item.context.contains('conflict') ||
        item.semanticCluster.contains('conflict')) {
      return 'conflict';
    }
    return null;
  }

  static void _insertExtra(
    List<FrequencyBehaviorV2Item> sequence,
    FrequencyBehaviorV2Item extra,
  ) {
    List<int>? bestKey;
    var bestPos = -1;
    for (var pos = 0; pos <= sequence.length; pos++) {
      if (_runLengthIfInsert(sequence, pos, extra) >
          FrequencyBehaviorV2Contract.maxConsecutiveSamePrimary) {
        continue;
      }
      final prev = pos == 0 ? null : sequence[pos - 1];
      final next = pos == sequence.length ? null : sequence[pos];
      final key = [
        _neighborDimOk(prev, extra) ? 1 : 0,
        _neighborDimOk(next, extra) ? 1 : 0,
        _neighborContextOk(prev, extra) ? 1 : 0,
        _neighborContextOk(next, extra) ? 1 : 0,
        _neighborClusterOk(prev, extra) ? 1 : 0,
        _neighborClusterOk(next, extra) ? 1 : 0,
      ];
      if (bestKey == null || _keyGreater(key, bestKey)) {
        bestKey = key;
        bestPos = pos;
      }
    }
    if (bestPos < 0) {
      throw StateError('no_valid_extra_insert_pos:${extra.itemId}');
    }
    sequence.insert(bestPos, extra);
  }

  static bool _neighborDimOk(
    FrequencyBehaviorV2Item? neighbor,
    FrequencyBehaviorV2Item extra,
  ) {
    if (neighbor == null) return true;
    return neighbor.primaryDimensions.single != extra.primaryDimensions.single;
  }

  static bool _neighborContextOk(
    FrequencyBehaviorV2Item? neighbor,
    FrequencyBehaviorV2Item extra,
  ) {
    if (neighbor == null) return true;
    return _contextFamily(neighbor) != _contextFamily(extra);
  }

  static bool _neighborClusterOk(
    FrequencyBehaviorV2Item? neighbor,
    FrequencyBehaviorV2Item extra,
  ) {
    if (neighbor == null) return true;
    return neighbor.semanticCluster != extra.semanticCluster;
  }

  static int _runLengthIfInsert(
    List<FrequencyBehaviorV2Item> sequence,
    int pos,
    FrequencyBehaviorV2Item extra,
  ) {
    final dim = extra.primaryDimensions.single;
    var run = 1;
    for (var i = pos - 1; i >= 0; i--) {
      if (sequence[i].primaryDimensions.single != dim) break;
      run++;
    }
    for (var i = pos; i < sequence.length; i++) {
      if (sequence[i].primaryDimensions.single != dim) break;
      run++;
    }
    return run;
  }

  static void _softenAdjacentFamilies(List<FrequencyBehaviorV2Item> sequence) {
    for (var i = 1; i < sequence.length; i++) {
      if (!_similarFamily(sequence[i - 1], sequence[i])) continue;
      for (var j = i + 1; j < sequence.length; j++) {
        _swap(sequence, i, j);
        final improved = _consecutivePrimaryOk(sequence) &&
            !_similarFamily(sequence[i - 1], sequence[i]);
        if (improved) break;
        _swap(sequence, i, j);
      }
    }
  }

  static bool _similarFamily(
    FrequencyBehaviorV2Item a,
    FrequencyBehaviorV2Item b,
  ) {
    if (a.semanticCluster == b.semanticCluster) return true;
    final ta = _thematicTag(a);
    final tb = _thematicTag(b);
    return ta != null && ta == tb;
  }

  static void _swap(List<FrequencyBehaviorV2Item> sequence, int i, int j) {
    final tmp = sequence[i];
    sequence[i] = sequence[j];
    sequence[j] = tmp;
  }

  static bool _consecutivePrimaryOk(List<FrequencyBehaviorV2Item> sequence) {
    var run = 1;
    for (var i = 1; i < sequence.length; i++) {
      if (sequence[i].primaryDimensions.single ==
          sequence[i - 1].primaryDimensions.single) {
        run++;
        if (run > FrequencyBehaviorV2Contract.maxConsecutiveSamePrimary) {
          return false;
        }
      } else {
        run = 1;
      }
    }
    return true;
  }

  bool _isSelectable(
    FrequencyBehaviorV2Item item,
    Map<String, Map<String, dynamic>> reviewByItemId,
  ) {
    if (item.primaryDimensions.length != 1) return false;
    if (!FrequencyBehaviorV2Contract.isCanonicalDimension(
      item.primaryDimensions.single,
    )) {
      return false;
    }
    final review = reviewByItemId[item.itemId];
    if (review != null) {
      if (review['selector_eligible'] != true) return false;
      if (review['rewrite_pending'] == true) return false;
      if (review['drop_from_selectable'] == true) return false;
      if (review['processing_style_present'] == true) return false;
      if (review['primary_review_pending'] == true) return false;
      final unresolved = review['unresolved_dimension_labels'];
      if (unresolved is List && unresolved.isNotEmpty) return false;
      final status = review['review_status']?.toString();
      if (status == 'manual_review' ||
          status == 'rewrite_pending' ||
          status == 'dropped_from_selectable') {
        return false;
      }
    }
    return item.options.every((o) => o.behavioralWeights.isNotEmpty);
  }
}

File: functions/src/frequency_behavior_v2_selector.js
'use strict';

const contract = require('./frequency_behavior_v2_contract');
const { FrequencyBehaviorV2Rng } = require('./frequency_behavior_v2_rng');

function isSelectable(item, review) {
  if (!item.primary_dimension) return false;
  if (!contract.isCanonicalDimension(item.primary_dimension)) return false;
  if (review) {
    if (review.selector_eligible !== true) return false;
    if (review.rewrite_pending === true) return false;
    if (review.drop_from_selectable === true) return false;
    if (review.processing_style_present === true) return false;
    if (review.primary_review_pending === true) return false;
    const unresolved = review.unresolved_dimension_labels;
    if (Array.isArray(unresolved) && unresolved.length > 0) return false;
    const status = review.review_status != null ? String(review.review_status) : null;
    if (
      status === 'manual_review' ||
      status === 'rewrite_pending' ||
      status === 'dropped_from_selectable'
    ) {
      return false;
    }
  }
  const optionIds = item.authored_option_ids || [];
  for (const oid of optionIds) {
    const opt = item.options[oid];
    if (!opt || !opt.behavioral_weights || Object.keys(opt.behavioral_weights).length === 0) {
      return false;
    }
  }
  return optionIds.length > 0;
}

function candidateRank({
  selectorVersion,
  bankVersion,
  sessionSeed,
  dimension,
  questionId,
}) {
  return FrequencyBehaviorV2Rng.fromParts([
    selectorVersion,
    bankVersion,
    sessionSeed,
    dimension,
    questionId,
  ]).nextUint32();
}

function nearDupIndex(clusters, eligibleIds) {
  const eligible = new Set(eligibleIds);
  const out = {};
  for (let i = 0; i < clusters.length; i++) {
    const members = clusters[i].filter((id) => eligible.has(id));
    if (members.length < 2) continue;
    for (const id of members) {
      if (out[id] === undefined) out[id] = i;
    }
  }
  return out;
}

function contextFamily(item) {
  if (!item.context || item.context.length === 0) return 'unclassified';
  return item.context[0];
}

function thematicTag(item) {
  const text = `${item.prompt || ''} ${item.semantic_cluster || ''}`.toLowerCase();
  if (/aile|anne|baba|kayın|çocuk|cocuk/.test(text)) return 'family';
  if (/para|fatura|hesap|ödeme|odeme|borç|borc/.test(text)) return 'payment';
  if (
    (item.context || []).includes('conflict') ||
    String(item.semantic_cluster || '').includes('conflict')
  ) {
    return 'conflict';
  }
  return null;
}

function keyGreater(a, b) {
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) return a[i] > b[i];
  }
  return false;
}

function neighborDimOk(neighbor, extra) {
  if (!neighbor) return true;
  return neighbor.primary_dimension !== extra.primary_dimension;
}

function neighborContextOk(neighbor, extra) {
  if (!neighbor) return true;
  return contextFamily(neighbor) !== contextFamily(extra);
}

function neighborClusterOk(neighbor, extra) {
  if (!neighbor) return true;
  return neighbor.semantic_cluster !== extra.semantic_cluster;
}

function runLengthIfInsert(sequence, pos, extra) {
  const dim = extra.primary_dimension;
  let run = 1;
  for (let i = pos - 1; i >= 0; i--) {
    if (sequence[i].primary_dimension !== dim) break;
    run++;
  }
  for (let i = pos; i < sequence.length; i++) {
    if (sequence[i].primary_dimension !== dim) break;
    run++;
  }
  return run;
}

function insertExtra(sequence, extra) {
  let bestKey = null;
  let bestPos = -1;
  for (let pos = 0; pos <= sequence.length; pos++) {
    if (
      runLengthIfInsert(sequence, pos, extra) >
      contract.MAX_CONSECUTIVE_SAME_PRIMARY
    ) {
      continue;
    }
    const prev = pos === 0 ? null : sequence[pos - 1];
    const next = pos === sequence.length ? null : sequence[pos];
    const key = [
      neighborDimOk(prev, extra) ? 1 : 0,
      neighborDimOk(next, extra) ? 1 : 0,
      neighborContextOk(prev, extra) ? 1 : 0,
      neighborContextOk(next, extra) ? 1 : 0,
      neighborClusterOk(prev, extra) ? 1 : 0,
      neighborClusterOk(next, extra) ? 1 : 0,
    ];
    if (bestKey === null || keyGreater(key, bestKey)) {
      bestKey = key;
      bestPos = pos;
    }
  }
  if (bestPos < 0) {
    throw new Error(`no_valid_extra_insert_pos:${extra.item_id}`);
  }
  sequence.splice(bestPos, 0, extra);
}

function similarFamily(a, b) {
  if (a.semantic_cluster === b.semantic_cluster) return true;
  const ta = thematicTag(a);
  const tb = thematicTag(b);
  return ta != null && ta === tb;
}

function swap(sequence, i, j) {
  const tmp = sequence[i];
  sequence[i] = sequence[j];
  sequence[j] = tmp;
}

function consecutivePrimaryOk(sequence) {
  let run = 1;
  for (let i = 1; i < sequence.length; i++) {
    if (sequence[i].primary_dimension === sequence[i - 1].primary_dimension) {
      run++;
      if (run > contract.MAX_CONSECUTIVE_SAME_PRIMARY) return false;
    } else {
      run = 1;
    }
  }
  return true;
}

function softenAdjacentFamilies(sequence) {
  for (let i = 1; i < sequence.length; i++) {
    if (!similarFamily(sequence[i - 1], sequence[i])) continue;
    for (let j = i + 1; j < sequence.length; j++) {
      swap(sequence, i, j);
      const improved =
        consecutivePrimaryOk(sequence) &&
        !similarFamily(sequence[i - 1], sequence[i]);
      if (improved) break;
      swap(sequence, i, j);
    }
  }
}

function composeManifest({
  bank,
  sessionSeed,
  sessionId,
  createdAt,
  excludeUnresolvedReview = true,
}) {
  const selectorVersion = contract.SELECTOR_VERSION;
  const bankVersion = bank.bank_version;
  const itemsById = bank.items_by_id;
  const reviewByItemId = bank.review_by_item_id || {};
  const nearDuplicateClusters = bank.near_duplicate_clusters || [];

  const eligible = [];
  for (const item of bank.items) {
    const review = reviewByItemId[item.item_id];
    if (excludeUnresolvedReview && !isSelectable(item, review)) continue;
    eligible.push(item);
  }

  const eligibleIds = eligible.map((i) => i.item_id);
  const ndIndex = nearDupIndex(nearDuplicateClusters, eligibleIds);
  const usedNearDup = new Set();

  const extraRng = FrequencyBehaviorV2Rng.forStream({
    selectorVersion,
    bankVersion,
    sessionSeed,
    stream: 'extra_slots',
  });
  const extraOrder = extraRng.shuffledCopy([...contract.CANONICAL_DIMENSIONS]);
  const extraDims = new Set([extraOrder[0], extraOrder[1]]);

  const pickedByDim = {};
  for (const dim of contract.CANONICAL_DIMENSIONS) {
    const quota = extraDims.has(dim)
      ? contract.SESSION_BASE_PER_DIMENSION + 1
      : contract.SESSION_BASE_PER_DIMENSION;
    const ranked = eligible.filter((item) => item.primary_dimension === dim);
    ranked.sort((a, b) => {
      const ra = candidateRank({
        selectorVersion,
        bankVersion,
        sessionSeed,
        dimension: dim,
        questionId: a.item_id,
      });
      const rb = candidateRank({
        selectorVersion,
        bankVersion,
        sessionSeed,
        dimension: dim,
        questionId: b.item_id,
      });
      if (ra !== rb) return ra - rb;
      return a.item_id < b.item_id ? -1 : a.item_id > b.item_id ? 1 : 0;
    });
    if (ranked.length < quota) {
      throw new Error(`insufficient_candidates:${dim} have=${ranked.length} need=${quota}`);
    }

    const picked = [];
    const clusterCounts = {};
    const pickedSet = new Set();

    function nearDupBlocked(c) {
      const nd = ndIndex[c.item_id];
      return nd != null && usedNearDup.has(nd);
    }

    function absorb(c) {
      pickedSet.add(c.item_id);
      picked.push(c);
      clusterCounts[c.semantic_cluster] = (clusterCounts[c.semantic_cluster] || 0) + 1;
      const nd = ndIndex[c.item_id];
      if (nd != null) usedNearDup.add(nd);
    }

    while (picked.length < quota) {
      const remaining = ranked.filter(
        (c) => !pickedSet.has(c.item_id) && !nearDupBlocked(c),
      );
      if (remaining.length === 0) {
        throw new Error(`underfilled:${dim} have=${picked.length} need=${quota}`);
      }
      let chosen = remaining[0];
      if ((clusterCounts[chosen.semantic_cluster] || 0) > 0) {
        const altLimit =
          contract.SOFT_CLUSTER_LOOKAHEAD < remaining.length - 1
            ? contract.SOFT_CLUSTER_LOOKAHEAD
            : remaining.length - 1;
        for (let i = 1; i <= altLimit; i++) {
          const alt = remaining[i];
          if ((clusterCounts[alt.semantic_cluster] || 0) === 0) {
            chosen = alt;
            break;
          }
        }
      }
      absorb(chosen);
    }
    pickedByDim[dim] = picked;
  }

  const queues = {};
  for (const dim of contract.CANONICAL_DIMENSIONS) {
    const qRng = FrequencyBehaviorV2Rng.forStream({
      selectorVersion,
      bankVersion,
      sessionSeed,
      stream: `queue|${dim}`,
    });
    queues[dim] = qRng.shuffledCopy(pickedByDim[dim]);
  }

  const interleaveRng = FrequencyBehaviorV2Rng.forStream({
    selectorVersion,
    bankVersion,
    sessionSeed,
    stream: 'interleave_order',
  });
  const dimOrder = interleaveRng.shuffledCopy([...contract.CANONICAL_DIMENSIONS]);

  const sequence = [];
  for (let round = 0; round < contract.SESSION_BASE_PER_DIMENSION; round++) {
    for (const dim of dimOrder) {
      sequence.push(queues[dim].shift());
    }
  }
  const extras = [];
  for (const dim of dimOrder) {
    extras.push(...queues[dim]);
  }
  for (const extra of extras) {
    insertExtra(sequence, extra);
  }
  softenAdjacentFamilies(sequence);

  if (sequence.length !== contract.SESSION_ITEM_COUNT) {
    throw new Error(`session_length_${sequence.length}`);
  }
  if (!consecutivePrimaryOk(sequence)) {
    throw new Error('consecutive_primary_violation');
  }

  const derivedId =
    sessionId ||
    `frequency_v2_${FrequencyBehaviorV2Rng.fnv1a32(
      `${selectorVersion}|${bankVersion}|${sessionSeed}`,
    )
      .toString(16)
      .padStart(8, '0')}`;

  const questions = [];
  for (let i = 0; i < sequence.length; i++) {
    const item = sequence[i];
    const optRng = FrequencyBehaviorV2Rng.forStream({
      selectorVersion,
      bankVersion,
      sessionSeed,
      stream: `options|${item.item_id}`,
    });
    const authored = item.authored_option_ids.slice();
    const presented = optRng.shuffledCopy(authored);
    questions.push({
      question_id: item.item_id,
      primary_dimension: item.primary_dimension,
      presentation_index: i,
      presented_option_order: presented,
    });
  }

  return {
    schema_version: contract.SESSION_MANIFEST_SCHEMA_VERSION,
    selector_version: selectorVersion,
    bank_version: bankVersion,
    session_id: derivedId,
    session_seed: sessionSeed,
    locale: bank.locale,
    created_at: createdAt || null,
    question_ids: questions.map((q) => q.question_id),
    questions,
    items_by_id: itemsById,
  };
}

module.exports = {
  composeManifest,
  isSelectable,
  thematicTag,
  contextFamily,
};
